Error querying association with basic collection types

JdbcSQLException occurred when try to query an association defined with Basic Collection Types

Having Student domain class, with the following body

class Student {
    String name
    String email

    static constraints = {
        name blank: false
        email blank: false, unique: true, email: true
    }

    static hasMany = [schools: String]

    static mapping = {
        table 'students'
        schools joinTable: [column: 'school']
    }
}

And from an StudentController action

    def index() {
        List<String> emailList = Student.createCriteria().list {
            'in'('schools', ['School2'])

            projections {
                property 'email'
            }
        }

        render emailList
    }

I am getting the following stacktrace

2017-06-21 16:22:23.146 ERROR --- [io-8080-exec-10] o.h.engine.jdbc.spi.SqlExceptionHelper   : Parameter "#1" is not set; SQL statement:
select this_.email as y0_ from students this_ where this_.id in (?) [90012-194]
2017-06-21 16:22:23.188 ERROR --- [io-8080-exec-10] o.g.web.errors.GrailsExceptionResolver   : JdbcSQLException occurred when processing request: [GET] /student/index
Parameter "#1" is not set; SQL statement:
select this_.email as y0_ from students this_ where this_.id in (?) [90012-194]. Stacktrace follows:

java.lang.reflect.InvocationTargetException: null
	at org.grails.core.DefaultGrailsControllerClass$ReflectionInvoker.invoke(DefaultGrailsControllerClass.java:211)
	at org.grails.core.DefaultGrailsControllerClass.invoke(DefaultGrailsControllerClass.java:188)
	at org.grails.web.mapping.mvc.UrlMappingsInfoHandlerAdapter.handle(UrlMappingsInfoHandlerAdapter.groovy:90)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:963)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:897)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:970)
	at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:861)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:846)
	at org.springframework.boot.web.filter.ApplicationContextHeaderFilter.doFilterInternal(ApplicationContextHeaderFilter.java:55)
	at org.grails.web.servlet.mvc.GrailsWebRequestFilter.doFilterInternal(GrailsWebRequestFilter.java:77)
	at org.grails.web.filters.HiddenHttpMethodFilter.doFilterInternal(HiddenHttpMethodFilter.java:67)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: org.hibernate.exception.GenericJDBCException: could not extract ResultSet
	at org.hibernate.exception.internal.StandardSQLExceptionConverter.convert(StandardSQLExceptionConverter.java:47)
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:111)
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:97)
	at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.extract(ResultSetReturnImpl.java:79)
	at org.hibernate.loader.Loader.getResultSet(Loader.java:2122)
	at org.hibernate.loader.Loader.executeQueryStatement(Loader.java:1905)
	at org.hibernate.loader.Loader.executeQueryStatement(Loader.java:1881)
	at org.hibernate.loader.Loader.doQuery(Loader.java:925)
	at org.hibernate.loader.Loader.doQueryAndInitializeNonLazyCollections(Loader.java:342)
	at org.hibernate.loader.Loader.doList(Loader.java:2622)
	at org.hibernate.loader.Loader.doList(Loader.java:2605)
	at org.hibernate.loader.Loader.listIgnoreQueryCache(Loader.java:2434)
	at org.hibernate.loader.Loader.list(Loader.java:2429)
	at org.hibernate.loader.criteria.CriteriaLoader.list(CriteriaLoader.java:109)
	at org.hibernate.internal.SessionImpl.list(SessionImpl.java:1787)
	at org.hibernate.internal.CriteriaImpl.list(CriteriaImpl.java:363)
	at org.grails.orm.hibernate.query.AbstractHibernateCriteriaBuilder.invokeMethod(AbstractHibernateCriteriaBuilder.java:1687)
	at org.grails.orm.hibernate.query.AbstractHibernateCriteriaBuilder.list(AbstractHibernateCriteriaBuilder.java:1558)
	at demo.app.StudentController.index(StudentController.groovy:6)
	... 14 common frames omitted
Caused by: org.h2.jdbc.JdbcSQLException: Parameter "#1" is not set; SQL statement:
select this_.email as y0_ from students this_ where this_.id in (?) [90012-194]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:345)
	at org.h2.message.DbException.get(DbException.java:179)
	at org.h2.message.DbException.get(DbException.java:155)
	at org.h2.expression.Parameter.checkSet(Parameter.java:81)
	at org.h2.command.Prepared.checkParameters(Prepared.java:164)
	at org.h2.command.CommandContainer.query(CommandContainer.java:112)
	at org.h2.command.Command.executeQuery(Command.java:201)
	at org.h2.jdbc.JdbcPreparedStatement.executeQuery(JdbcPreparedStatement.java:110)
	at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.extract(ResultSetReturnImpl.java:70)
	... 29 common frames omitted

### Steps to Reproduce

1. Clone this repo https://github.com/ilmoralito/Error-querying-basic-collection-types
2. Run the repo
3. Go to this location http://localhost:8080/student/index

### Expected Behaviour

Should print a list of student emails

### Actual Behaviour

Error 500: Internal Server Error

URI /student/index
Class org.h2.jdbc.JdbcSQLException
Message null
Caused by Parameter "#1" is not set; SQL statement: select this_.email as y0_ from students this_ where this_.id in (?) [90012-194]

### Environment Information

- **Operating System**: Ubuntu
- **Grails Version:** Using grails version 3.2.9
- **JDK Version:** Using java version 8u121

### Example Application

- TODO: https://github.com/ilmoralito/Error-querying-basic-collection-types

--- a/grails-data-hibernate5/core/src/main/groovy/grails/orm/HibernateCriteriaBuilder.java
+++ b/grails-data-hibernate5/core/src/main/groovy/grails/orm/HibernateCriteriaBuilder.java
@@ -145,6 +145,7 @@ public class HibernateCriteriaBuilder extends AbstractHibernateCriteriaBuilder {
      * @see #createAlias(String, String)
      */
     public Criteria createAlias(String associationPath, String alias, int joinType) {
+        aliasMap.put(associationPath, alias);
         return criteria.createAlias(associationPath, alias, JoinType.parse(joinType));
     }
 

--- a/grails-data-hibernate5/core/src/main/groovy/org/grails/orm/hibernate/query/AbstractHibernateCriteriaBuilder.java
+++ b/grails-data-hibernate5/core/src/main/groovy/org/grails/orm/hibernate/query/AbstractHibernateCriteriaBuilder.java
@@ -621,6 +621,7 @@ public abstract class AbstractHibernateCriteriaBuilder extends GroovyObjectSuppo
      * @throws org.hibernate.HibernateException Indicates a problem creating the sub criteria
      */
     public Criteria createAlias(String associationPath, String alias) {
+        aliasMap.put(associationPath, alias);
         return criteria.createAlias(associationPath, alias);
     }
 

--- a/grails-data-hibernate5/core/src/main/groovy/org/grails/orm/hibernate/query/AbstractHibernateCriteriaBuilder.java
+++ b/grails-data-hibernate5/core/src/main/groovy/org/grails/orm/hibernate/query/AbstractHibernateCriteriaBuilder.java
@@ -72,6 +72,9 @@ import org.grails.datastore.mapping.query.Query;
 import org.grails.datastore.mapping.query.api.BuildableCriteria;
 import org.grails.datastore.mapping.query.api.QueryableCriteria;
 import org.grails.datastore.mapping.reflect.NameUtils;
+import org.grails.datastore.mapping.model.PersistentEntity;
+import org.grails.datastore.mapping.model.PersistentProperty;
+import org.grails.datastore.mapping.model.types.Basic;
 import org.grails.orm.hibernate.AbstractHibernateDatastore;
 
 /**
@@ -1286,7 +1289,17 @@ public abstract class AbstractHibernateCriteriaBuilder extends GroovyObjectSuppo
         if (values instanceof List) {
             values = convertArgumentList((List) values);
         }
-        addToCriteria(Restrictions.in(propertyName, values == null ? Collections.EMPTY_LIST : values));
+
+        // Handle basic collection types (hasMany to String/Integer/etc.)
+        // These are stored in a separate join table and cannot use simple Restrictions.in().
+        // Instead, create an alias to the collection table and restrict on 'elements'.
+        if (isBasicCollectionProperty(propertyName)) {
+            String alias = propertyName + ALIAS;
+            createAlias(propertyName, alias);
+            addToCriteria(Restrictions.in(alias + ".elements", values == null ? Collections.EMPTY_LIST : values));
+        } else {
+            addToCriteria(Restrictions.in(propertyName, values == null ? Collections.EMPTY_LIST : values));
+        }
         return this;
     }
 
@@ -1331,7 +1344,15 @@ public abstract class AbstractHibernateCriteriaBuilder extends GroovyObjectSuppo
         }
 
         propertyName = calculatePropertyName(propertyName);
-        addToCriteria(Restrictions.in(propertyName, values));
+
+        // Handle basic collection types (hasMany to String/Integer/etc.)
+        if (isBasicCollectionProperty(propertyName)) {
+            String alias = propertyName + ALIAS;
+            createAlias(propertyName, alias);
+            addToCriteria(Restrictions.in(alias + ".elements", values));
+        } else {
+            addToCriteria(Restrictions.in(propertyName, values));
+        }
         return this;
     }
 
@@ -1981,6 +2002,23 @@ public abstract class AbstractHibernateCriteriaBuilder extends GroovyObjectSuppo
         else {
             c.addOrder(ignoreCase ? Order.asc(sort).ignoreCase() : Order.asc(sort));
         }
+    }
+
+    /**
+     * Checks if the given property name refers to a Basic collection type
+     * (e.g. hasMany: [schools: String]). Basic collections are stored in
+     * a separate join table and require special handling for 'in' queries.
+     */
+    private boolean isBasicCollectionProperty(String propertyName) {
+        if (datastore == null) {
+            return false;
+        }
+        PersistentEntity entity = datastore.getMappingContext().getPersistentEntity(targetClass.getName());
+        if (entity == null) {
+            return false;
+        }
+        PersistentProperty property = entity.getPropertyByName(propertyName);
+        return property instanceof Basic;
     }
 
     /**

--- a/grails-data-hibernate5/core/src/main/groovy/org/grails/orm/hibernate/query/AbstractHibernateCriteriaBuilder.java
+++ b/grails-data-hibernate5/core/src/main/groovy/org/grails/orm/hibernate/query/AbstractHibernateCriteriaBuilder.java
@@ -67,14 +67,14 @@ import org.springframework.beans.BeanUtils;
 import org.springframework.core.convert.ConversionService;
 
 import grails.gorm.MultiTenant;
+import org.grails.datastore.mapping.model.PersistentEntity;
+import org.grails.datastore.mapping.model.PersistentProperty;
+import org.grails.datastore.mapping.model.types.Basic;
 import org.grails.datastore.mapping.multitenancy.MultiTenancySettings;
 import org.grails.datastore.mapping.query.Query;
 import org.grails.datastore.mapping.query.api.BuildableCriteria;
 import org.grails.datastore.mapping.query.api.QueryableCriteria;
 import org.grails.datastore.mapping.reflect.NameUtils;
-import org.grails.datastore.mapping.model.PersistentEntity;
-import org.grails.datastore.mapping.model.PersistentProperty;
-import org.grails.datastore.mapping.model.types.Basic;
 import org.grails.orm.hibernate.AbstractHibernateDatastore;
 
 /**
@@ -1284,6 +1284,9 @@ public abstract class AbstractHibernateCriteriaBuilder extends GroovyObjectSuppo
                     propertyName + "] and values [" + values + "] not allowed here."));
         }
 
+        // Preserve the original property name before alias prefix is applied,
+        // since isBasicCollectionProperty needs the raw property name for entity lookup.
+        String originalPropertyName = propertyName;
         propertyName = calculatePropertyName(propertyName);
 
         if (values instanceof List) {
@@ -1293,9 +1296,15 @@ public abstract class AbstractHibernateCriteriaBuilder extends GroovyObjectSuppo
         // Handle basic collection types (hasMany to String/Integer/etc.)
         // These are stored in a separate join table and cannot use simple Restrictions.in().
         // Instead, create an alias to the collection table and restrict on 'elements'.
-        if (isBasicCollectionProperty(propertyName)) {
-            String alias = propertyName + ALIAS;
-            createAlias(propertyName, alias);
+        if (isBasicCollectionProperty(originalPropertyName)) {
+            String alias;
+            if (aliasMap.containsKey(propertyName)) {
+                alias = aliasMap.get(propertyName);
+            } else {
+                alias = propertyName + ALIAS;
+                createAlias(propertyName, alias);
+                aliasMap.put(propertyName, alias);
+            }
             addToCriteria(Restrictions.in(alias + ".elements", values == null ? Collections.EMPTY_LIST : values));
         } else {
             addToCriteria(Restrictions.in(propertyName, values == null ? Collections.EMPTY_LIST : values));
@@ -1343,12 +1352,21 @@ public abstract class AbstractHibernateCriteriaBuilder extends GroovyObjectSuppo
                     propertyName + "] and values [" + values + "] not allowed here."));
         }
 
+        // Preserve the original property name before alias prefix is applied,
+        // since isBasicCollectionProperty needs the raw property name for entity lookup.
+        String originalPropertyName = propertyName;
         propertyName = calculatePropertyName(propertyName);
 
         // Handle basic collection types (hasMany to String/Integer/etc.)
-        if (isBasicCollectionProperty(propertyName)) {
-            String alias = propertyName + ALIAS;
-            createAlias(propertyName, alias);
+        if (isBasicCollectionProperty(originalPropertyName)) {
+            String alias;
+            if (aliasMap.containsKey(propertyName)) {
+                alias = aliasMap.get(propertyName);
+            } else {
+                alias = propertyName + ALIAS;
+                createAlias(propertyName, alias);
+                aliasMap.put(propertyName, alias);
+            }
             addToCriteria(Restrictions.in(alias + ".elements", values));
         } else {
             addToCriteria(Restrictions.in(propertyName, values));

--- a/grails-data-hibernate5/core/src/test/groovy/grails/gorm/tests/BasicCollectionInQuerySpec.groovy
+++ b/grails-data-hibernate5/core/src/test/groovy/grails/gorm/tests/BasicCollectionInQuerySpec.groovy
@@ -123,6 +123,34 @@ class BasicCollectionInQuerySpec extends Specification {
         then: "all matching students are found (duplicates possible from OR on join table)"
         results.unique().sort() == ['dave@test.com', 'eve@test.com', 'frank@test.com']
     }
+
+    def "in query on basic collection with pre-existing alias should reuse it"() {
+        given:
+        def s1 = new BcStudent(name: "Grace", email: "grace@test.com")
+        s1.addToSchools("Yale")
+        s1.addToSchools("Princeton")
+        s1.save()
+
+        def s2 = new BcStudent(name: "Hank", email: "hank@test.com")
+        s2.addToSchools("Princeton")
+        s2.save()
+
+        def s3 = new BcStudent(name: "Ivy", email: "ivy@test.com")
+        s3.addToSchools("Columbia")
+        s3.save(flush: true)
+
+        when: "an alias is explicitly created before using in() with the raw property name"
+        def results = BcStudent.createCriteria().list {
+            createAlias("schools", "sch")
+            'in'('schools', ['Princeton'])
+            projections {
+                property 'email'
+            }
+        }
+
+        then: "the existing alias is reused instead of creating a duplicate"
+        results.sort() == ['grace@test.com', 'hank@test.com']
+    }
 }
 
 @Entity

--- a/grails-data-hibernate5/core/src/test/groovy/grails/gorm/tests/BasicCollectionInQuerySpec.groovy
+++ b/grails-data-hibernate5/core/src/test/groovy/grails/gorm/tests/BasicCollectionInQuerySpec.groovy
@@ -91,6 +91,38 @@ class BasicCollectionInQuerySpec extends Specification {
         then:
         results.sort() == ['alice2@test.com', 'bob2@test.com']
     }
+
+    def "multiple in queries on same basic collection should not fail with duplicate alias"() {
+        given:
+        def s1 = new BcStudent(name: "Dave", email: "dave@test.com")
+        s1.addToSchools("MIT")
+        s1.addToSchools("Harvard")
+        s1.save()
+
+        def s2 = new BcStudent(name: "Eve", email: "eve@test.com")
+        s2.addToSchools("Stanford")
+        s2.addToSchools("Berkeley")
+        s2.save()
+
+        def s3 = new BcStudent(name: "Frank", email: "frank@test.com")
+        s3.addToSchools("MIT")
+        s3.addToSchools("Stanford")
+        s3.save(flush: true)
+
+        when:
+        def results = BcStudent.createCriteria().list {
+            or {
+                'in'('schools', ['MIT'])
+                'in'('schools', ['Stanford'])
+            }
+            projections {
+                property 'email'
+            }
+        }
+
+        then: "all matching students are found (duplicates possible from OR on join table)"
+        results.unique().sort() == ['dave@test.com', 'eve@test.com', 'frank@test.com']
+    }
 }
 
 @Entity

--- a/grails-data-hibernate5/core/src/test/groovy/grails/gorm/tests/BasicCollectionInQuerySpec.groovy
+++ b/grails-data-hibernate5/core/src/test/groovy/grails/gorm/tests/BasicCollectionInQuerySpec.groovy
@@ -1,0 +1,107 @@
+/*
+ *  Licensed to the Apache Software Foundation (ASF) under one
+ *  or more contributor license agreements.  See the NOTICE file
+ *  distributed with this work for additional information
+ *  regarding copyright ownership.  The ASF licenses this file
+ *  to you under the Apache License, Version 2.0 (the
+ *  "License"); you may not use this file except in compliance
+ *  with the License.  You may obtain a copy of the License at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing,
+ *  software distributed under the License is distributed on an
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied.  See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ */
+package grails.gorm.tests
+
+import grails.gorm.annotation.Entity
+import grails.gorm.transactions.Rollback
+import org.grails.orm.hibernate.HibernateDatastore
+import spock.lang.AutoCleanup
+import spock.lang.Issue
+import spock.lang.Shared
+import spock.lang.Specification
+
+/**
+ * Reproduces https://github.com/apache/grails-core/issues/14610
+ *
+ * When a domain has hasMany to a basic type (String), using 'in' on
+ * that collection property in criteria queries fails with
+ * "Parameter #1 is not set".
+ */
+@Rollback
+class BasicCollectionInQuerySpec extends Specification {
+
+    @Shared @AutoCleanup HibernateDatastore datastore =
+        new HibernateDatastore(BcStudent)
+
+    @Issue("https://github.com/apache/grails-core/issues/14610")
+    def "in query on basic collection type should work"() {
+        given:
+        def s1 = new BcStudent(name: "Alice", email: "alice@test.com")
+        s1.addToSchools("School1")
+        s1.addToSchools("School2")
+        s1.save()
+
+        def s2 = new BcStudent(name: "Bob", email: "bob@test.com")
+        s2.addToSchools("School2")
+        s2.addToSchools("School3")
+        s2.save()
+
+        def s3 = new BcStudent(name: "Charlie", email: "charlie@test.com")
+        s3.addToSchools("School3")
+        s3.save(flush: true)
+
+        when:
+        def results = BcStudent.createCriteria().list {
+            'in'('schools', ['School2'])
+            projections {
+                property 'email'
+            }
+        }
+
+        then:
+        results.sort() == ['alice@test.com', 'bob@test.com']
+    }
+
+    def "workaround using createAlias on basic collection"() {
+        given:
+        def s1 = new BcStudent(name: "Alice2", email: "alice2@test.com")
+        s1.addToSchools("SchoolA")
+        s1.addToSchools("SchoolB")
+        s1.save()
+
+        def s2 = new BcStudent(name: "Bob2", email: "bob2@test.com")
+        s2.addToSchools("SchoolB")
+        s2.save(flush: true)
+
+        when:
+        def results = BcStudent.createCriteria().list {
+            createAlias("schools", "s")
+            'in'("s.elements", ["SchoolB"])
+            projections {
+                property 'email'
+            }
+        }
+
+        then:
+        results.sort() == ['alice2@test.com', 'bob2@test.com']
+    }
+}
+
+@Entity
+class BcStudent {
+    String name
+    String email
+
+    static hasMany = [schools: String]
+
+    static constraints = {
+        name blank: false
+        email blank: false
+    }
+}

